### PR TITLE
fixes an issue in C++ DQN related to choosing illegal actions

### PR DESCRIPTION
**Pull request by sabify**
Fixes an issue related to the poorly implementation of the algorithm that causes choosing illegal actions.

### CODE REVIEW

File: open_spiel/algorithms/dqn_torch/dqn.cc [hunk 1]
```diff
@@ -141,37 +142,37 @@ void DQN::AddTransition(const State& prev_state,
 }
 
 Action DQN::EpsilonGreedy(std::vector<float> info_state,
-                          std::vector<Action> legal_actions,
-                          double epsilon) {
+                          std::vector<Action> legal_actions, double epsilon) {
   Action action;
   if (legal_actions.empty()) {
     // In some simultaneous games, some players can have no legal actions.
-    return 0;
```
Comment (lanctot) [inline open_spiel/algorithms/dqn_torch/dqn.cc:149]: This should be zero actually (see other comment)

Comment (sabify) [reply open_spiel/algorithms/dqn_torch/dqn.cc]: Done

**Comment by sabify** [comment]
Done.

**Comment by lanctot** [comment]
@christianjans @michalsustr can you take a quick look at this PR?

I'm not very familiar with the PyTorch C++ API.

@sabify can you explain in words what the problem was. I don't quite understand.

**Comment by sabify** [comment]
Short answer: Integer/float overflows in addition can make illegal actions, legal. The masking was not implemented correctly.

Long answer: I'm going to explain it with an example. 

We have 5 total actions (`num_actions_`) and 2 legal actions. The current algorithm first try to allocate a tensor of 5 elements that has 3 elements with illegal action value (-1e9) and 2 elements for legal actions with zero values. Then, it is going to sum this tensor with Q values before selecting the highest value. The issue is that if we have a very large negative value and sum it with illegal action value, it can cause negative integer/float overflow or a big Q value in illegal action's indexes can be big enough that the addition be bigger than the legal action's values and selecting the highest value is no longer in the legal actions.

This PR correctly masks the legal values with torch built-in functions.

**Comment by lanctot** [comment]
That makes sense... wow, very subtle... nice catch!!

@Asugawara @jhtschultz

**Comment by Asugawara** [comment]
Thank you for fixing the bug!

>it can cause negative integer/float overflow or a big Q value in illegal action's indexes can be big enough that the addition be bigger than the legal action's values and selecting the highest value is no longer in the legal actions.

Am I correct in understanding that the target value is much smaller than 'kIllegalActionLogitsPenalty', so the possibility is very small?

p.s. Is that the same??
https://github.com/deepmind/open_spiel/blob/b3dae9dcd34fa75171baac09831e60d17c95ce1d/open_spiel/algorithms/dqn_torch/dqn.cc#L233-L238

**Comment by lanctot** [comment]
@sabify I have already imported this before your last commit. So when I do a sync on Monday, it will be merged without your most recent commit. You can submit a new PR to make those changes.

**Comment by sabify** [comment]
> p.s. Is that the same??

Yes, and I just updated the file with new implementation.

> Am I correct in understanding that the target value is much smaller than 'kIllegalActionLogitsPenalty', so the possibility is very small?

It maybe depends on the game and its dqn configuration. In my testing for a 4-player game with 53 actions, 2822 tensor size and 128x128 hidden layers configuration, the issue arise at 100000th game and the game would never proceed anymore. Unfortunately, the games in c++ dqn example won't catch this issue.

**Comment by sabify** [comment]
> @sabify I have already imported this before your last commit. So when I do a sync on Monday, it will be merged without your most recent commit. You can submit a new PR to make those newest changes. (After it is merged on Monday.)

I will create new PR and they don't have any conflict, you can import both...

**Comment by lanctot** [comment]
That works, thanks.

**Pull request closed without merging**